### ISSUE TITLE
Some weirdness in reported data

### ISSUE THREAD

**Issue by kmansoft** [VictoriaMetrics/VictoriaMetrics#120]
Hi,

I just switched my project from InfluxDB to Victoria and greatly enjoying Victoria's performance.

However, I'm seeing something weird in the data (query results) reported by Victoria.

The project is a simple monitoring system for Linux - CPU usage, disk space, web server requests, that kind of thing. It's still in development, so I only run it (two processes - the server and the collector) during certain times when I'm awake so it's not on 24/7 yet.

- First, Victoria seems to report some data for time periods where I know that no data was collected (submitted). First attachment, I've highlighted "wrong" areas, the OK area is when data collection was started.

![image](https://user-images.githubusercontent.com/1686219/61821122-2fe25900-ae5f-11e9-809d-ba5ec98a2471.png)

- Second, Victoria seems to lose some data sometimes, this is the second attachment. In general, all series are submitted at the same time, as part of a single batch every time, so this means that all charts should be in sync with respect to "have data" / "no data". This however is not happening.

![image](https://user-images.githubusercontent.com/1686219/61821142-3f61a200-ae5f-11e9-8fda-8203dd2c79c2.png)

- I didn't see anything like this with Influx. Now I've hit a performance problem with Influx and so Victoria's performance is really a big deal to me. But server monitoring is a kind of thing where you want to be able to trust your data, and so far Victoria seems to have a problem with this.

How do you even debug this? Let's say you were interested in fixing this - I'd need to submit "something" so you could reproduce the issues, how to do that if Victoria's data store is a whole bunch of files?

**Comment by kmansoft** [comment]
PS - I'm running current code from git.

**Comment by valyala** [comment]
Let's try debugging the issue.

VictoriaMetrics follows [Prometheus data model](https://prometheus.io/docs/prometheus/latest/querying/basics/#gotchas) when queried with PromQL:

> When queries are run, timestamps at which to sample data are selected independently of the actual present time series data. This is mainly to support cases like aggregation (sum, avg, and so on), where multiple aggregated time series do not exactly align in time. Because of their independence, Prometheus needs to assign a value at those timestamps for each relevant time series. It does so by simply taking the newest sample before this timestamp.

This means that the data returned via `/api/v1/query` or `/api/v1/query_range` is _calculated_ according to the rules mentioned above.

> How do you even debug this?

It is possible to extract raw data via [/api/v1/export](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/README.md#how-to-export-time-series) API. Debugging must start with obtaining the raw data via `/api/v1/export` API and comparing to to the ingested data. Note that timestamps in InfluxDB line protocol [are in _nanoseconds_ by default](https://docs.influxdata.com/influxdb/v1.7/write_protocols/line_protocol_tutorial/#timestamp), while VictoriaMetrics stores and exports timestamps in _milliseconds_ via `/api/v1/export` API.

So the first debugging steps are:
1. Insert data via Influx line protocol
2. Obtain it via `/api/v1/export`
3. Compare the inserted data with the obtained data

**Comment by kmansoft** [comment]
Thanks @kmansoft 

Here is another case.

![image](https://user-images.githubusercontent.com/1686219/61842935-99d02200-aea2-11e9-966c-12fa0dff5cb6.png)

**Comment by kmansoft** [comment]
Bingo:

```
curl -G 'http://localhost:8428/api/v1/query_range' --data-urlencode 'query=mem_value{node="56f202a6-1248-4efe-8549-0daebe93f0b0",sub="swap_used"}' --data-urlencode 'start=1564022353' --data-urlencode 'end=1564024182' --data-urlencode 'step=1m' |  python -m json.tool 
```

returns

```
{
    "data": {
        "result": [
            {
                "metric": {
                    "__name__": "mem_value",
                    "db": "clearview",
                    "node": "56f202a6-1248-4efe-8549-0daebe93f0b0",
                    "sub": "swap_used"
                },
                "values": [
                    [
                        1564022353,
                        "NaN"
                    ],
                    [
                        1564022413,
                        "NaN"
                    ],
                    [
                        1564022473,
                        "NaN"
                    ],
                    [
                        1564022533,
                        "NaN"
                    ],
                    [
                        1564022593,
                        "NaN"
                    ],
                    [
                        1564022653,
                        "0"
                    ],
                    [
                        1564022713,
                        "NaN"
                    ],
                    [
                        1564022773,
                        "NaN"
                    ],
                    [
                        1564022833,
                        "NaN"
                    ],
                    [
                        1564022893,
                        "NaN"
                    ],
                    [
                        1564022953,
                        "NaN"
                    ],
                    [
                        1564023013,
                        "NaN"
                    ],
                    [
                        1564023073,
                        "0"
                    ],
                    [
                        1564023133,
                        "NaN"
                    ],
                    [
                        1564023193,
                        "NaN"
                    ],
                    [
                        1564023253,
                        "NaN"
                    ],
                    [
                        1564023313,
                        "NaN"
                    ],
                    [
                        1564023373,
                        "NaN"
                    ],
                    [
                        1564023433,
                        "NaN"
                    ],
                    [
                        1564023493,
                        "0"
                    ],
                    [
                        1564023553,
                        "NaN"
                    ],
                    [
                        1564023613,
                        "NaN"
                    ],
                    [
                        1564023673,
                        "NaN"
                    ],
                    [
                        1564023733,
                        "NaN"
                    ],
                    [
                        1564023793,
                        "NaN"
                    ],
                    [
                        1564023853,
                        "NaN"
                    ],
                    [
                        1564023913,
                        "0"
                    ],
                    [
                        1564023973,
                        "0"
                    ],
                    [
                        1564024033,
                        "0"
                    ],
                    [
                        1564024093,
                        "0"
                    ],
                    [
                        1564024153,
                        "0"
                    ]
                ]
            }
        ],
        "resultType": "matrix"
    },
    "status": "success"
}
```

In this case there is one "stray" value but it's there:

```
                        1564022653,
                        "0"
```

The values at 1564023913 and newer is good, I woke up, booted the computer, and started the software.

But the value at 1564022653 should be a NaN - at that time I was sleeping, with the computer turned off.

I then stopped victoria and took a snapshot of the data directory. It is attached. After a restart the issue persists.

PS - my server process submits data in order, without specifying time values at all.
[victoria-metrics-issue-120.tar.gz](https://github.com/VictoriaMetrics/VictoriaMetrics/files/3429593/victoria-metrics-issue-120.tar.gz)

**Comment by kmansoft** [comment]
Export:

```
curl -G 'http://localhost:8428/api/v1/export' --data-urlencode 'match[]=mem_value{node="56f202a6-1248-4efe-8549-0daebe93f0b0",sub="swap_used"}' --data-urlencode 'start=1564022353' --data-urlencode 'end=1564024182' --data-urlencode 'step=1m' |  python -m json.tool 

{
    "metric": {
        "__name__": "mem_value",
        "db": "clearview",
        "node": "56f202a6-1248-4efe-8549-0daebe93f0b0",
        "sub": "swap_used"
    },
    "timestamps": [
        1564022608238,
        1564023025892,
        1564023443546,
        1564023861200,
        1564023866367,
        1564023871454,
        1564023876541,
        1564023881628,
        1564023886715,
        1564023891802,
        1564023896889,
        1564023901976,
        1564023907063,
        1564023912150,
        1564023917237,
        1564023922324,
        1564023927411,
        1564023932498,
        1564023937585,
        1564023942672,
        1564023947759,
        1564023952846,
        1564023957933,
        1564023963020,
        1564023968107,
        1564023973194,
        1564023978275,
        1564023983372,
        1564023988469,
        1564023993566,
        1564023998663,
        1564024003760,
        1564024008857,
        1564024013954,
        1564024019051,
        1564024024148,
        1564024029245,
        1564024034342,
        1564024039439,
        1564024044536,
        1564024049633,
        1564024054730,
        1564024059876,
        1564024064978,
        1564024070052
    ],
    "values": [
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0
    ]
}
```

The all zeros is also wrong, I'd say there should be NaNs (probably not possible in JSON) or the values from more than ~15 minutes back should not be there at all.

**Comment by kmansoft** [comment]
One more case. A different "page" of my software, different (just created) series.

So I see this literally all the time.

![image](https://user-images.githubusercontent.com/1686219/61853568-e2003c00-aec4-11e9-94f7-1a75e0b81f0d.png)

**Comment by valyala** [comment]
Raw data from `/api/v1/export` for the metric `mem_value{node="56f202a6-1248-4efe-8549-0daebe93f0b0",sub="swap_used"}` contains the following timestamps in milliseconds:

```
        1564022608238,
        1564023025892,
        1564023443546,
        1564023861200,
        1564023866367,
        1564023871454,
```

They correspond to the following UTC times:

```
2019-07-25T02:43:28
2019-07-25T02:50:25
2019-07-25T02:57:23
2019-07-25T03:04:21
2019-07-25T03:04:26
2019-07-25T03:04:31
```

The interval between the first three timestamps is ~7 minutes, while the interval between the remaining timestamp is 5 seconds. Probably this is related to monotonic time handling in the Operating System during sleep and after wake up. Go [uses monotonic time](https://golang.org/pkg/time/).

Do these invalid timestamps appear only after computer wake up from sleep? Or they appear when the computer is on for extended periods of time?

**Comment by kmansoft** [comment]
- I do not use suspend, at all. The computer is fully turned off and then on.

- It seems too much (or too little) for jitter.

- Both the "good" and the "bad" series are submitted together, using a single batch (InfluxDB API) every time.

- It happens over and over again if I start the software after a break.

- My submit code does not specify timestamps, letting InfluxDB or Victoria use the current time by default.

**Comment by valyala** [comment]
> It happens over and over again if I start the software after a break.

Most likely the issue is related to VictoriaMetrics restart.
Additionally we have the following facts:
- invalid data points appear on time ranges when VictoriaMetrics wasn't running
- interval between invalid data points is ~7 minutes
- interval between valid data points is 5 seconds

Try stopping VictoriaMetrics, then deleting the folder `<-storageDataPath>/cache`, then starting VictoriaMetrics again and checking whether invalid data points remain. This should rule out possible issues with response caching.

**Comment by kmansoft** [comment]
> Most likely the issue is related to VictoriaMetrics restart.

I'm not restarting VictoriaMetrics.

Let me clarify:

- I don't use suspend or hibernate so the computer is off at night

- Victoria Metrics is set to start via systemd when the computer boots. I don't stop or kill VictoriaMetrics.

- Now *my* software is started manually - it's in development - and I may do this several times during a day and/or it may not be running at all during some time periods.

- When I do start my software, it starts feeding data into already running VictoriaMetrics, there is an agent part (which collects some data and writes it into Victoria using Influx API) and a reporting part (which queries data).

- As data starts showing after a restart of *my* software, I pretty often see that some charts are out of sync with others, and there are data points which were never written / sent to Victoria because *my* software was not running.

I will try deleting the cache next time I see it.

**Comment by kmansoft** [comment]
> Try stopping VictoriaMetrics, then deleting the folder <-storageDataPath>/cache, then starting VictoriaMetrics again and checking whether invalid data points remain. This should rule out possible issues with response caching.

Yes they do remain after: 1) stop victoria 2) delete cache folder 3) restart victoria.

![image](https://user-images.githubusercontent.com/1686219/61875017-6321f800-aef2-11e9-8507-d76dbd219c53.png)

**Comment by kmansoft** [comment]
And here is what it looks like with InfluxDB, note that all charts are in sync and only have data on the right side, the most recent minute or two, and everything before that is empty indicating no data (as opposed to some values, 0 or not).

![image](https://user-images.githubusercontent.com/1686219/61875369-23a7db80-aef3-11e9-98a9-5c11ea6a1eea.png)

**Comment by valyala** [comment]
Which queries does your software send to VictoriaMetrics? VictoriaMetrics follows Prometheus' data model when queried via `/api/v1/query` or `/api/v1/query_range` handlers. In this model data points are calculated at regular interval, which is equal to `step` value passed to these handlers. Each data point value is calculated as the closest real (raw) value before the given timestamp for the data point.
InfluxDB has different data model for querying - it always returns real (raw) values from queries and doesn't calculate additional values at regular interval. The most closest results to InfluxDB's must be obtained via `/api/v1/export` handler, which returns real (raw) values from the database.

I investigated [the attached data](https://github.com/VictoriaMetrics/VictoriaMetrics/files/3429593/victoria-metrics-issue-120.tar.gz) and it looks like your software continues sending data to VictoriaMetrics at lower rate after you think it is stopped. This results in graphs with data on time ranges when the software wasn't running. Certain metrics such as `disk_value` and `cpu_value` aren't sent when the software is stopped, while other metrics such as `mem_value`, `net_value`, `sys_value` continue arriving when the software must be stopped. For example, `mem_value` metrics are sent every 5 seconds when the software is running and continue arriving at different rates when the software must be stopped. `mem_value{sub="swap_free"}` and `mem_value{sub="swap_used"}` was sent every 12 seconds during the first day, then the interval increased to 36 seconds and later it has been increased to 400 seconds. Other `mem_value` metrics such as `{sub="real_free"}`, `{sub="real_used"}`, `{sub="real_cache"}` and `{sub="real_buffers"}` are sent only when the software was running.

This hypothesis can be verified by monitoring `vm_http_requests_total{path="/write", protocol="influx"}` and `vm_rows_inserted_total{type="influx"}` metrics exposed at `/metrics` page of VictoriaMetrics service. These metrics must remain constant when the software doesn't run. The easiest way to verify this is to run Prometheus that scrapes the `/metrics` page of VictoriaMetrics and then building graphs for the metrics mentioned above. Graphs must increase when software was running and must remain flat (constant) when the software wasn't running.

**Comment by kmansoft** [comment]
> Which queries does your software send to VictoriaMetrics?

Writes: InfluxDB API with *no* time specified (so it's assigned by the time DB, and can't go backwards or be wrong). A single batch every 5 seconds.

Reads: "query" and "query_range".

> looks like your software continues sending data to VictoriaMetrics at lower rate after you think it is stopped

Sorry no. Today's new case:

- The computer was turned off last night around 23:30. Not suspended or hibernated. Turned off.

- I booted the computer at 6:20 am, Victoria was started from systemd

- Started my software manually (no system) at 6:21

- Waited maybe a minute

- You can see the results below

- Maybe there is a bug in InfluxDB API? Since you found actual data in the database (i.e. it's on the storage side not on the querying side)?

I can do the prometheus diags you outlined above, but since it's linked to the theory about "software still sending data when it should not be", I haven't yet.

![image](https://user-images.githubusercontent.com/1686219/61924071-e719c580-af6e-11e9-9adc-c88df34fd583.png)

![image](https://user-images.githubusercontent.com/1686219/61924075-ed0fa680-af6e-11e9-9fc4-049559758bcf.png)

![image](https://user-images.githubusercontent.com/1686219/61924080-f26cf100-af6e-11e9-8d1c-afb69c435b83.png)

[victoria-metrics.tar.gz](https://github.com/VictoriaMetrics/VictoriaMetrics/files/3434062/victoria-metrics.tar.gz)

**Comment by valyala** [comment]
> Maybe there is a bug in InfluxDB API? Since you found actual data in the database (i.e. it's on the storage side not on the querying side)?

I'm investigating this case, but without any findings yet :(

> I booted the computer at 6:20 am, Victoria was started from systemd.

And graphs clearly show data points before 6:20 when VictoriaMetrics wasn't running because the computer where it runs was turned off. This suggests the following case: invalid data points could be added there in a short period of time by a third-party client, who sent Influx line protocol data with explicit regular, but invalid timestamps. By default VictoriaMetrics stores data points with timestamps up to 2 days in the future. So invalid data points with timestamps in the future may be stored by a third-party client, and now they clutter graphs. In order to verify this theory, the following steps must be performed:
   - Stopping VictoriaMetrics
   - Removing contents of the `-storageDataPath` directory, so all the possible invalid data with timestamps in the future is deleted.
   - Starting VictoriaMetrics with new `-httpListenAddr` port in order to prevent from writing invalid data from third-party clients.
   - Starting your software with the updated TCP port for writing into VictoriaMetrics.
   - Verifying whether invalid datapoints exist

> I can do the prometheus diags you outlined above, but since it's linked to the theory about "software still sending data when it should not be", I haven't yet.

Please verify this by monitoring `vm_http_requests_total{path="/write", protocol="influx"}` and `vm_rows_inserted_total{type="influx"}` metrics exposed at `/metrics` page of VictoriaMetrics service.

**Comment by kmansoft** [comment]
Ummm you mean some client which sends data with data points in the future?

No not that either.

Early on in my daily sequence I start Victoria and my software and I can
see in the charts that there are no data points. Then they start showing -
both good and bad.

Besides, what kind of software would find Victoria on a nonstandard port
number and exactly match my series names which include a sha1 hash (client
ID)?

If such clever software exists, I want it to write my code for me :)

пт, 26 июл. 2019 г., 13:56 Aliaksandr Valialkin <notifications@github.com>:

**Comment by valyala** [comment]
OK, then it would be great if you could minimize the test case showing the issue with invalid data. Could you strip all the irrelevant code from your software leaving only the minimal client part that sends data to VictoriaMetrics? There is no need in sending real data or all the data. It is enough sending fake data for a couple of metrics where VictoriaMetrics shows invalid data. For instance, `mem_value{sub="swap_free"}`.

Share the resulting minimal client code here that reproduces the issue. This will greatly simplify finding the root cause of the issue and fixing it.

**Comment by kmansoft** [comment]
> OK, then it would be great if you could minimize the test case showing the issue with invalid data

Right, I'll work on this.

**Comment by kmansoft** [comment]
The software also needs Mongo and has a config file. Hope this is not a problem.

**Comment by valyala** [comment]
> The software also needs Mongo and has a config file. Hope this is not a problem.

It would be better removing Mongo from the minimal client. The client must only send data to VM in Influx line protocol. It shouldn't query VM, since the issue is on the data ingestion path.

**Comment by kmansoft** [comment]
It's a heisenbug too - just now I was not able to make it happen across
three restarts.

пт, 26 июл. 2019 г., 20:27 Aliaksandr Valialkin <notifications@github.com>:

**Comment by jcmoraisjr** [comment]
Hi, I can confirm the following behavior which could be related. Below are two grafana screenshots from the same metric, the only difference is the start time which will use a different zoom.

8h30 - 11h30:

![8h](https://user-images.githubusercontent.com/1457820/62400604-31460c80-b556-11e9-98c6-a877490236b5.png)

9h30 - 11h30:

![9h](https://user-images.githubusercontent.com/1457820/62400634-428f1900-b556-11e9-8fd9-402643baef2a.png)

But the machine with prometheus - which is the only one that stores on VictoriaMetrics - and with VictoriaMetrics itself was turned on only about 10h55.

VictoriaMetrics deployment is just a single container. Is there any useful info from my environment which would help?

**Comment by valyala** [comment]
@jcmoraisjr , thanks for details. Your setup is slightly different than @kmansoft setup above - you write data to VictoriaMetrics via Prometheus remote_write API, while @kmansoft writes data via Influx line protocol. This rules out the initial suspicion with Influx line protocol parsing bug.

Could you provide metric name with all the labels for the wrong time series?
Is there only a single wrong time series or there are other wrong time series? Could you provide screenshots for them with their names and all the labels?

**Comment by jcmoraisjr** [comment]
Moving to #141

**Comment by valyala** [comment]
Closing this issue, since it has been moved to #141

**Comment by valyala** [comment]
The issue must be fixed in [v1.24.1](https://github.com/VictoriaMetrics/VictoriaMetrics/releases).
